FancyBatching.md

```diff
@@ -0,0 +1,505 @@
+    expose the tuning of these parameters.
+
+    - `private ResponseClassifier RetryClassifier { get; set; }`  This is the
+    type we use in Azure.Core to customize Retry behavior.  It will basically
+    retry on 422, 409, 503 status codes and the standard IOException and
+    RequestFailedException with status code 0.
+
+    - TODO: If we end up using PID Controllers for control stabilization, should
+    we expose their gain configuration?
+
+## Using SearchIndexingBufferedSender
+
+- `public class SearchIndexingBufferedSender<T>`
+    - Add Indexing Actions
+        - `void DeleteDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task DeleteDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void IndexDocuments(IndexDocumentsBatch<T>, CancellationToken)`
+        - `Task IndexDocumentsAsync(IndexDocumentsBatch<T>, CancellationToken)`
+
+            Note these are using a batch which we already have conveniences for
+            building up.
+
+        - `void MergeDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task MergeDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void MergeOrUploadDocuments(IEnumerable, CancellationToken)`
+        - `Task MergeOrUploadDocumentsAsync(IEnumerable<T>, CancellationToken)`
+        - `void UploadDocuments(IEnumerable<T>, CancellationToken)`
+        - `Task UploadDocumentsAsync(IEnumerable<T>, CancellationToken)`
+
+        We're not returning any data from these operations.  They may submit
+        something immediately or we might be waiting for a flush.
+
+        The provided cancellation token will be used for any requests sent, any
+        notifications raised, etc.  If a flush happens when the timer elapses,
+        then we'll use the `AutoFlushIntervalCancellationToken` instead.  A 
+        customer could reuse a token or link them together for fine grained
+        control.
+
+        We're also not including the `IndexDocumentsOptions` parameter because
+        that currently only serves to configure exception behavior which we're
+        doing exclusively through events.
+
+        - Delete by key
+
+            Add customization for "delete with just a key" scenarios for dynamic
+            `SearchDocument`.  We're using extension methods in .NET but other
+            languages might need to figure out a different approach.
+
+            - `public static class SearchExtensions`
+                - `static void DeleteDocuments(this SearchIndexingBufferedSender<SearchDocument>, string, IEnumerable<string>, CancellationToken)`
+                - `static Task DeleteDocumentsAsync(this SearchIndexingBufferedSender<SearchDocument>, string, IEnumerable<string>, CancellationToken)`
+
+    - `void Flush(CancellationToken)` and `Task FlushAsync(CancellationToken)`
+        will manually flush any pending operations.  This could result in
+        multiple requests.  If a request is already in progress, whatever's left
```

Comment:
I'm intrigued on manual flush potentially leading to multiple requests. Are the multiple requests because of possible retrying of documents? Does it include the scenario where more documents are added to the sender while the flush is processing then sending another request because we've exceeded batch size again? I would worry about getting a customer stuck in a stack where they have a thread `await`ing until this loop breaks. For the background timer based flush I am fine with it because we don't have an applications main thread on lock.

Reply:
We have to support it for manual flushing like:

```C#
options.AutoFlush = false;
// ...
indexer.UploadDocuments(allMsftEmployees);
indxer.Flush();
```

But yes, that also covers the scenario where more documents are added while a publisher is submitting.  I'm less worried about the edge case where somebody calls `Flush` and then 10 other folks have documents slowly trickling in.

One other possibility would be spawning a separate publisher thread in the background and `UploadDocuments` or `Flush` never waiting through one or more requests.